ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/MLSdkAsyncHttpResponseHandler.java

```diff
@@ -0,0 +1,239 @@
+        this.parameters = parameters;
+        this.tensorOutputs = tensorOutputs;
+        this.connector = connector;
+        this.scriptService = scriptService;
+        this.mlGuard = mlGuard;
+    }
+
+    @Override
+    public void onHeaders(SdkHttpResponse response) {
+        SdkHttpFullResponse sdkResponse = (SdkHttpFullResponse) response;
+        log.debug("received response headers: " + sdkResponse.headers());
+        this.statusCode = sdkResponse.statusCode();
+    }
+
+    @Override
+    public void onStream(Publisher<ByteBuffer> stream) {
+        stream.subscribe(new MLResponseSubscriber());
+    }
+
+    @Override
+    public void onError(Throwable error) {
+        log.error(error.getMessage(), error);
+        if (statusCode == null) {
+            actionListener
+                .onFailure(
+                    new OpenSearchStatusException(
+                        "Error on communication with remote model: " + error.getMessage(),
+                        RestStatus.INTERNAL_SERVER_ERROR
+                    )
+                );
+        } else {
+            actionListener
+                .onFailure(
+                    new OpenSearchStatusException(
+                        "Error on communication with remote model: " + error.getMessage(),
+                        RestStatus.fromCode(statusCode)
+                    )
+                );
+        }
+    }
+
+    private void processResponse(
+        Integer statusCode,
+        String body,
+        Map<String, String> parameters,
+        Map<Integer, ModelTensors> tensorOutputs
+    ) {
+        if (Strings.isBlank(body)) {
+            log.error("Remote model response body is empty!");
+            actionListener.onFailure(new OpenSearchStatusException("No response from model", RestStatus.BAD_REQUEST));
+        } else {
+            if (statusCode < HttpStatus.SC_OK || statusCode > HttpStatus.SC_MULTIPLE_CHOICES) {
+                log.error("Remote server returned error code: {}", statusCode);
+                actionListener.onFailure(new OpenSearchStatusException(REMOTE_SERVICE_ERROR + body, RestStatus.fromCode(statusCode)));
+            } else {
+                try {
+                    ModelTensors tensors = processOutput(body, connector, scriptService, parameters, mlGuard);
+                    tensors.setStatusCode(statusCode);
+                    tensorOutputs.put(countDownLatch.getSequence(), tensors);
+                } catch (Exception e) {
+                    log.error("Failed to process response body: {}", body, e);
+                    actionListener.onFailure(new MLException("Fail to execute predict in aws connector", e));
+                }
+            }
+        }
+    }
+
+    private void reOrderTensorResponses(Map<Integer, ModelTensors> tensorOutputs) {
+        List<ModelTensors> modelTensors = new ArrayList<>();
+        TreeMap<Integer, ModelTensors> sortedMap = new TreeMap<>(tensorOutputs);
```

Comment:
Seems not necessary to create a new TreeMap for every predict request. Can we use such way to iterate `sortedMap` with sorted order 
```
for (int i = 1; i<=sortedMap.size(); i++) {
    ModelTensors tensors = sortedMap.get(i);
}
```